flocker/common/_retry.py

```diff
@@ -275,3 +279,178 @@ def should_retry(e):
     should_retry.wait_secs = retry_wait
 
     return retry(effect, should_retry)
+
+
+_TRY_UNTIL_SUCCESS = u"flocker:failure-retry"
+_TRY_RETRYING = _TRY_UNTIL_SUCCESS + u":retrying"
+_TRY_FAILURE = _TRY_UNTIL_SUCCESS + u":failure"
+_TRY_SUCCESS = _TRY_UNTIL_SUCCESS + u":success"
+
+
+def get_default_retry_steps(
+    delay=timedelta(seconds=0.1),
+    max_time=timedelta(minutes=2)
+):
+    """
+    Retry every 0.1 seconds for 2 minutes.
+    """
+    repetitions = max_time.total_seconds() / delay.total_seconds()
+    return repeat(delay, int(repetitions))
+
+
+def retry_always(exc_type, value, traceback):
+    pass
+
+
+def retry_if(predicate):
+    """
+    Create a predicate compatible with ``with_retry``
+    which will retry if the raised exception satisfies the given predicate.
+
+    :param predicate: A one-argument callable which will be called with the
+        raised exception instance only.  It should return ``True`` if a retry
+        should be attempted, ``False`` otherwise.
+    """
+    def should_retry(exc_type, value, traceback):
+        if predicate(value):
+            return None
+        raise exc_type, value, traceback
+    return should_retry
+
+
+def decorate_methods(obj, decorator):
+    """
+    Return a wrapper around ``obj`` with ``decorator`` applied to all of its
+    method calls.
+
+    For example, to retry on IOError up to 5 times with a 3 second delay
+    between each try::
+
+        retry_three_times = partial(
+            with_retry,
+            should_retry=retry_if(lambda exc: isinstance(exc, IOError)),
+            steps=[timedelta(seconds=3)] * 5,
+        )
+        obj_with_three_retries = decorate_methods(obj, retry_three_times)
+
+    :param callable decorator: A unary callable that takes a method and returns
+        a method.
+
+    :return: An object like ``obj`` but with all the methods decorated.
+    """
+    return _DecoratedInstance(obj, decorator)
```

Comment:
:question: This should probably be moved to a different module, since it's not really about retries.